src/AssembleElemSolverAlgorithm.C

```diff
@@ -66,6 +68,80 @@ AssembleElemSolverAlgorithm::initialize_connectivity()
   eqSystem_->linsys_->buildElemToNodeGraph(partVec_);
 }
 
+static constexpr int simdLen = stk::simd::ndoubles;
+
+int
+calculate_shared_mem_bytes_per_thread(int lhsSize, int rhsSize, int scratchIdsSize, int nDim,
+                                      ElemDataRequests& dataNeededByKernels)
+{
+    int bytes_per_thread = (rhsSize + lhsSize)*sizeof(double) + (2*scratchIdsSize)*sizeof(int) +
+                           get_num_bytes_pre_req_data<double>(dataNeededByKernels, nDim);
+    bytes_per_thread *= 2*simdLen;
+    return bytes_per_thread;
+}
+
+template<typename T>
+void set_zero(T* values, unsigned length)
+{
+    for(unsigned i=0; i<length; ++i) {
+        values[i] = 0;
+    }
+}
+
+int get_next_num_elems_simd(int bktIndex, int bktLength)
+{
+  int numElems = simdLen;
+  if (bktLength - bktIndex*simdLen < simdLen) {
+    numElems = bktLength - bktIndex*simdLen;
+  }
+  if (numElems < 0 || numElems > simdLen) {
+    std::cout<<"ERROR, simdElems="<<numElems<<" shouldn't happen!!"<<std::endl;
+    numElems = 0;
+  }
+  return numElems;
+}
+
+size_t get_simd_bucket_length(size_t bktLength)
+{
+    size_t simdBucketLen = bktLength/simdLen;
+    const size_t remainder = bktLength%simdLen;
+    if (remainder > 0) {
+      simdBucketLen += 1;
+    }
+    return simdBucketLen;
+}
+
+struct SharedMemData {
```

Comment:
Is the thinking that ScratchViews will only be used in ElemSolverAlgorithm? If not, should this be promoted to a header file?